lang/funcs/collection.go

```diff
@@ -800,7 +800,13 @@ var MatchkeysFunc = function.New(&function.Spec{
 		},
 	},
 	Type: func(args []cty.Value) (cty.Type, error) {
-		if !args[1].Type().Equals(args[2].Type()) {
+		argTypes := make([]cty.Type, 2)
+		for i := 0; i < 2; i++ {
+			argTypes[i] = args[i+1].Type()
+		}
+
+		ty, _ := convert.UnifyUnsafe(argTypes)
```

Comment:
I think we may need to repeat this step in `Impl` below and then convert `args[1]` and `args[2]` to the unified type, because the main loop in `Impl` is calling `stdlib.Equal` for each pair of elements and that will return false if the given element types are not the same.

This wouldn't come up in the situation that was causing the confusing error here, because the `Impl` function doesn't run at all if any of the arguments are unknown, but I think it would arise in the following situation:

```hcl
matchkeys(["a", "b"], [1, 2], [1, "2"])
```

In the above case, argument 1 would become a list of number while argument 2 would become a list of string. `convert.UnifyUnsafe` here would then unify that to list of string, but if we don't repeat the unify in `Impl` below then we'll end up comparing numbers to strings using `Equal`, and so that'll always return false.

The above example seems contrived because it's built out of constant values, but I expect stuff like that will crop up in real-world configurations where these lists are being constructed dynamically from values coming from variables and resource attributes.

Reply:
Oh odd, I thought I was doing that already - I've gone back and forth a few times so I must've reverted that when I wasn't looking 😄

Reply:
Excellent, I was also able to add a test for this scenario - I was briefly confused that I couldn't put a tuple in the test but coffee kicked in and I remembered this situation is exactly why we now have functions_test.go